### DISCUSSION TITLE
Does `instance_index` work on Metal ?

### DISCUSSION THREAD

**Discussion by olofwalker** [gfx-rs/wgpu#4928] [General]
I have rewritten my rendering pipeline to use `DrawIndexedIndirect` and rely on `instance_index` being set to `base_instance` to correctly index arrays with transforms and other stuff in the shader.

However, I can't get it to work, it seems like the builtin `instance_index` is always zero.

This is how the vertex shader entrypoint looks like
```
fn main_vs(
    @builtin(instance_index) object_id: u32,
    input: VertexInput,
) -> VertexOutput {
```

Here is how I create the indirect commands

```
        for (i, (vertex_count, indices_count)) in lengths.iter().enumerate() {
            indirect_data.push(DrawIndexedIndirect {
                vertex_count: *indices_count as u32, // Will be renamed in 0.19
                instance_count: 1,
                base_index: index_offset,
                vertex_offset,
                base_instance: i as u32,
            });
            vertex_offset += *vertex_count as i32;
            index_offset += *indices_count as u32;
        }
```

And yes, I have enabled the non-zero based instance index feature

```
        let (device, queue) = adapter
            .request_device(
                &DeviceDescriptor {
                    label: None,
                    features: Features::TEXTURE_BINDING_ARRAY
                        | Features::SAMPLED_TEXTURE_AND_STORAGE_BUFFER_ARRAY_NON_UNIFORM_INDEXING
                        | Features::PUSH_CONSTANTS
                        | Features::POLYGON_MODE_LINE
                        | Features::INDIRECT_FIRST_INSTANCE
                        | Features::VERTEX_WRITABLE_STORAGE,
                    limits: Limits {
                        max_push_constant_size: 4,
                        ..Limits::default()
                    },
                },
                None, // Trace path
            )
            .await
            .unwrap();
```

Is anyone else seeing this?

Specs: Macos Sonoma, Rust 1.72, WGPU 0.18

**Comment by cwfitzgerald** [comment]
It should, could you locally run our test for this thing. 

```rust
cargo install cargo-nextest
// in wgpu checkout
cargo xtask test vertex_indices
```

**Comment by olofwalker** [comment]
@cwfitzgerald Thanks, this works, and I quickly discovered the problem when running my code. 

My buffer holding `DrawIndexedIndirect` contains more than one entry, and I just figured that `draw_indexed_indirect` would draw all of them, but I should have used `multi_draw_indexed_indirect`.

The question is, why is there a need for two different functions doing in essences the same thing?

**Reply by cwfitzgerald** [reply]
It's a hardware limitation fundamentally. DX11 and GL4.0-4.2 only have single indirect. Multi draw indirect came in dx12/GL4.3. We emulate multi-draw on platforms that don't support it through repeated calls to draw_indirect.
